interval.hh

```diff
@@ -28,6 +28,22 @@
 #include <boost/range/algorithm/copy.hpp>
 #include <boost/range/adaptor/sliced.hpp>
 #include <boost/range/adaptor/transformed.hpp>
+#include <compare>
+
+inline
+bool
+require_ordering_and_on_equal_return(
+        std::strong_ordering input,
+        std::strong_ordering match_to_return_true,
+        bool return_if_equal) {
+    if (input == match_to_return_true) {
```

Comment:
Maybe `return input == match_to_return_true || (return_if_equal && input == std::strong_ordering::equal)` ?

That would avoid branches so no risk of branch misprediction.

Reply:
Whether you use if or logical operations says nothing about whether the compiler uses branches or not. Logically speaking they're exactly equivalent since the compiler is required not to evaluate the right hand side of && if the left hand side was false (or the right hand size of || if the left hand side was true), so a naive compiler would generate the same code. In practice optimal compilers notice the right hand side has no side effects, so they can convert both the if variant and the logical operator variant to arithmetic.

Reply:
And in fact they generate similar code:

https://godbolt.org/z/oYxW15

Reply:
And clang generates better code for the first version! Of course you can't predict that in advance.

https://godbolt.org/z/oYxW15